mne/viz/utils.py

```diff
@@ -1032,3 +1032,71 @@ def _plot_sensors(pos, colors, ch_names, title, show_names, show):
     fig.suptitle(title)
     plt_show(show)
     return fig
+
+
+def compute_scalings(scalings, inst):
+    """Compute scalings for each channel type automatically.
+
+    Parameters
+    ----------
+    scalings : dict
+        The scalings for each channel type. If any values are
+        'auto', this will automatically compute a reasonable
+        scaling for that channel type. Any values that aren't
+        'auto' will not be changed.
+    inst : instance of Raw or Epochs
+        The data for which you want to compute scalings. If Epochs,
+        data will be preloaded. If Raw, a window of max 20
+        seconds will be loaded from the middle of the data for
+        estimating scaling.
+
+    Returns
+    -------
+    scalings : dict
+        A scalings dictionary with updated values
+    """
+    from copy import deepcopy
+    from mne.io.base import _BaseRaw
+    from mne.io.pick import _picks_by_type
+    from mne.epochs import _BaseEpochs
+    if not isinstance(inst, (_BaseRaw, _BaseEpochs)):
+        raise ValueError('Must supply either Raw or Epochs')
+    if scalings is None:
+        # If scalings is None just return it and do nothing
+        return scalings
+
+    ch_types = _picks_by_type(inst.info)
+    unique_ch_types = [i_type[0] for i_type in ch_types]
+    if scalings == 'auto':
+        # If we want to auto-compute everything
+        scalings = dict((i_type, 'auto') for i_type in unique_ch_types)
+    if not isinstance(scalings, dict):
+        raise ValueError('scalings must be a dictionary of ch_type: val pairs')
+    scalings = deepcopy(scalings)
+
+    if inst.preload is False:
+        if isinstance(inst, _BaseRaw):
+            # Load a sample of data
+            n_secs = 20.
+            time_middle = np.mean(inst.times)
+            tmin = np.clip(time_middle - n_secs / 2., inst.times.min(), None)
+            tmax = np.clip(time_middle + n_secs / 2., None, inst.times.max())
+            data = inst._read_segment(tmin, tmax)
+        elif isinstance(inst, _BaseEpochs):
+            inst.load_data()
+    else:
+        data = inst._data
+    if isinstance(inst, _BaseEpochs):
+        data = np.hstack(inst._data)
+
+    # Iterate through ch types and update scaling if ' auto'
+    for key, value in scalings.items():
+        if value != 'auto':
+            continue
+        if key not in unique_ch_types:
+            raise ValueError("Sensor {0} doesn't exist in data".format(key))
+        this_ixs = [i_type[1] for i_type in ch_types if i_type[0] == key][0]
+        this_data = data[this_ixs]
+        scale_factor = np.percentile(np.abs(this_data.ravel()), 99.5)
```

Comment:
the `np.abs` will also cause a copy. Instead, you could do `np.max(np.percentile(this_data.ravel(), [0.5, 99.5]) * [-1, 1])`